llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp

```diff
@@ -784,6 +792,148 @@ createLinalgStrategyRemoveMarkersPass() {
+    if (!tileAndPosMapping.count(dim)) {
+      paddedShape.push_back(size);
+      continue;
+    }
+
+    Optional<int64_t> tileSize =
+        getConstantIntValue(tileAndPosMapping.lookup(dim));
+    assert((!inputType.isDynamicDim(dim) && tileSize.hasValue()) &&
+           "something goes really wrong...");
+    int64_t sizeWithPad = llvm::alignTo(size, tileSize.getValue());
+    paddedShape.push_back(sizeWithPad);
+  }
+  auto resultType =
+      RankedTensorType::get(paddedShape, inputType.getElementType());
+  return tensor::createPadHighOp(resultType, input, packOp.getPaddingValue(),
+                                 /*nofold=*/false, loc, builder);
+}
+
+/// Rewrites iree_linalg_ext.pack to tensor.pad + rank-reduced linalg.generic
+/// (transpose) ops.
+struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
+  using OpRewritePattern<PackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(PackOp packOp,
+                                PatternRewriter &rewriter) const final {
+    if (!packOp.hasTensorSemantics()) {
+      return rewriter.notifyMatchFailure(packOp, "require tensor semantics");
+    }
+
+    // The expand_shape op can be avoided if outer dimensions of result are all
+    // 1s. This can be relaxed if needed. A tensor.expand_shape will be
+    // generated in that case.
+    int64_t inputRank = packOp.getInputRank();
+    if (llvm::any_of(packOp.getOutputShape().take_front(inputRank),
+                     [](int64_t val) { return val != 1; })) {
+      return rewriter.notifyMatchFailure(
+          packOp, "require the outer dimension of the result are all 1s");
+    }
+
+    Value input = getInputOrPaddedInput(rewriter, packOp);
+
+    SmallVector<AffineExpr> inputExprs;
+    for (int64_t dim = 0; dim < inputRank; ++dim) {
+      inputExprs.push_back(rewriter.getAffineDimExpr(dim));
+    }
+    if (auto outerDims = packOp.getOuterDimsPerm()) {
+      inputExprs = interchange<AffineExpr>(inputExprs,
+                                           extractFromI64ArrayAttr(outerDims));
+    }
+    for (auto en :
+         llvm::enumerate(extractFromI64ArrayAttr(packOp.getInnerDimsPos()))) {
+      inputExprs[en.value()] =
+          rewriter.getAffineDimExpr(inputRank + en.index());
+    }
+
+    Location loc = packOp.getLoc();
+    auto inputType = input.getType().cast<RankedTensorType>();
+    auto nloops = packOp.getOutputRank();
+
+    Value empty = rewriter.create<tensor::EmptyOp>(loc, packOp.getOutputShape(),
+                                                   inputType.getElementType());
+    SmallVector<StringRef, 4> loopAttributeTypes(nloops,
+                                                 getParallelIteratorTypeName());
+    SmallVector<AffineMap, 2> indexingMaps = {
+        AffineMap::get(nloops, 0, inputExprs, rewriter.getContext()),
+        AffineMap::getMultiDimIdentityMap(nloops, rewriter.getContext())};
+
+    auto transposedOp = rewriter.create<linalg::GenericOp>(
```

Comment:
We're not able to use linalg.transpose because it requires the rank to be the same. Here we need a transpose op that expanding the shape (with unit dims).